Bug in _termvectors with artificial document?

I am using the `_termvectors` API with an artificial document, and I hit an inconsistent behavior, depending on which shard the request hits. Below I attach reproducing steps, but let me first explain what I do: I create an index with two shards and index one document with one field and the value "one one". When I request the term vectors of that document with `term_statistics`, I get `ttf=2` as expected. At this point I know that one shard contains the document, and the other shard has no documents at all.

When I submit a `_termvectors` request with an artificial document with the field `text:one`, then if the request hits the shard that holds the indexed document, I get back `ttf=2` which is expected. But if it hits the shard without any documents, it returns `ttf=1` which is confusing.

More so, if I send an artificial document with the field `text:two` (NOTE: two does not exist in any of the shards!), then the response is similar -- if the query hits the shard without any documents, it returns `ttf=1`, while if it hits the shard with the one indexed document, it does not return `ttf` at all (which is expected).

Here are the steps to reproduce:

**Create the index**
```curl -XPUT 'localhost:9200/tv_bug' -d'
{
  "mappings": {
    "doc": {
      "properties": {
        "text" : {
          "type": "text"
        }
      }
    }
  },
  "settings" : {
    "index" : {
      "number_of_shards" : 2,
      "number_of_replicas" : 0
    }
  }
}'
```

**Index one document**
```
curl -XPUT 'http://localhost:9200/tv_bug/doc/1?pretty=true' -d '{
  "text" : "one one"
}'
```

**Verify the document exists in one shard**
```
curl -XGET 'http://localhost:9200/tv_bug/doc/_search?preference=_shards:0&pretty=true'
curl -XGET 'http://localhost:9200/tv_bug/doc/_search?preference=_shards:1&pretty=true'
```
One of those requests will return 0 hits, the other 1 (in my case shard 0 did not return the hit)

**Get the Term Vectors of that one document**
```
curl -XGET 'http://localhost:9200/tv_bug/doc/1/_termvectors?pretty=true' -d '{
  "fields": ["text"],
  "term_statistics" : true
}'
```

**Get the TV of 'one' using an artificial document**

***From shard 0***
```
curl -XGET 'http://localhost:9200/tv_bug/doc/_termvectors?preference=_shards:0&pretty=true' -d '{
  "term_statistics" : true,
  "doc" : {
    "text" : "one"
  }
}'
```

***From shard 1***
```
curl -XGET 'http://localhost:9200/tv_bug/doc/_termvectors?preference=_shards:1&pretty=true' -d '{
  "term_statistics" : true,
  "doc" : {
    "text" : "one"
  }
}'
```

In my case, shard 0 returns `ttf=1` and shard 1 returns `ttf=2` (expected). Also, if you send the word "two" instead, you will see that shard 1 does not return `ttf` at all (expected), but shard 0 returns `ttf=1`.

BTW, and this may be a different bug report, when I send the TV requests with the `preference` parameter, I sometimes receive this error:

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "remote_transport_exception",
        "reason" : "[C-N1Zbn][127.0.0.1:9300][indices:data/read/tv[s]]"
      }
    ],
    "type" : "null_pointer_exception",
    "reason" : null
  },
  "status" : 500
}
```

Submitting the request multiple times eventually succeeds.

Here's the NPE stacktrace from the console BTW:

```
[2016-12-01T11:59:13,208][WARN ][r.suppressed             ] path: /tv_bug/doc/_termvectors, params: {pretty=true, preference=_shards:1, index=tv_bug, type=doc}
org.elasticsearch.transport.RemoteTransportException: [C-N1Zbn][127.0.0.1:9300][indices:data/read/tv[s]]
Caused by: java.lang.NullPointerException
        at org.elasticsearch.action.termvectors.TransportTermVectorsAction.shardOperation(TransportTermVectorsAction.java:79) ~[elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.action.termvectors.TransportTermVectorsAction.shardOperation(TransportTermVectorsAction.java:42) ~[elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$ShardTransportHandler.messageReceived(TransportSingleShardAction.java:293) ~[elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$ShardTransportHandler.messageReceived(TransportSingleShardAction.java:286) ~[elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33) ~[elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.transport.TransportService$6.doRun(TransportService.java:548) [elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:520) [elasticsearch-5.0.1.jar:5.0.1]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.0.1.jar:5.0.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_92]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_92]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_92]
```

--- a/core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsWriter.java
+++ b/core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsWriter.java
@@ -71,7 +71,7 @@ final class TermVectorsWriter {
 
             // if no terms found, take the retrieved term vector fields for stats
             if (topLevelTerms == null) {
-                topLevelTerms = fieldTermVector;
+                topLevelTerms = EMPTY_TERMS;
             }
 
             TermsEnum topLevelIterator = topLevelTerms.iterator();
@@ -292,4 +292,30 @@ final class TermVectorsWriter {
         // further...
         output.writeVLong(Math.max(0, value + 1));
     }
+
+    /** Implements an empty {@link Terms}. */
+    private static final Terms EMPTY_TERMS = new Terms() {
+        @Override public TermsEnum iterator() throws IOException { return EMPTY_TERMS_ENUM; }
+        @Override public long size() throws IOException { return 0; }
+        @Override public long getSumTotalTermFreq() throws IOException { return 0; }
+        @Override public long getSumDocFreq() throws IOException { return 0; }
+        @Override public int getDocCount() throws IOException { return 0; }
+        @Override public boolean hasFreqs() { return false; }
+        @Override public boolean hasOffsets() { return false; }
+        @Override public boolean hasPositions() { return false; }
+        @Override public boolean hasPayloads() { return false; }
+    };
+
+    /** Implements an empty {@link TermsEnum}. */
+    private static final TermsEnum EMPTY_TERMS_ENUM = new TermsEnum() {
+        @Override public BytesRef next() throws IOException { return null; }
+        @Override public SeekStatus seekCeil(BytesRef text) throws IOException { return SeekStatus.END; }
+        @Override public void seekExact(long ord) throws IOException { }
+        @Override public BytesRef term() throws IOException { return null; }
+        @Override public long ord() throws IOException { return 0; }
+        @Override public int docFreq() throws IOException { return 0; }
+        @Override public long totalTermFreq() throws IOException { return 0; }
+        @Override public PostingsEnum postings(PostingsEnum reuse, int flags) throws IOException { return null; }
+    };
+
 }

--- a/core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsWriter.java
+++ b/core/src/main/java/org/elasticsearch/action/termvectors/TermVectorsWriter.java
@@ -295,7 +295,7 @@ final class TermVectorsWriter {
 
     /** Implements an empty {@link Terms}. */
     private static final Terms EMPTY_TERMS = new Terms() {
-        @Override public TermsEnum iterator() throws IOException { return EMPTY_TERMS_ENUM; }
+        @Override public TermsEnum iterator() throws IOException { return TermsEnum.EMPTY; }
         @Override public long size() throws IOException { return 0; }
         @Override public long getSumTotalTermFreq() throws IOException { return 0; }
         @Override public long getSumDocFreq() throws IOException { return 0; }
@@ -304,18 +304,6 @@ final class TermVectorsWriter {
         @Override public boolean hasOffsets() { return false; }
         @Override public boolean hasPositions() { return false; }
         @Override public boolean hasPayloads() { return false; }
-    };
-
-    /** Implements an empty {@link TermsEnum}. */
-    private static final TermsEnum EMPTY_TERMS_ENUM = new TermsEnum() {
-        @Override public BytesRef next() throws IOException { return null; }
-        @Override public SeekStatus seekCeil(BytesRef text) throws IOException { return SeekStatus.END; }
-        @Override public void seekExact(long ord) throws IOException { }
-        @Override public BytesRef term() throws IOException { return null; }
-        @Override public long ord() throws IOException { return 0; }
-        @Override public int docFreq() throws IOException { return 0; }
-        @Override public long totalTermFreq() throws IOException { return 0; }
-        @Override public PostingsEnum postings(PostingsEnum reuse, int flags) throws IOException { return null; }
     };
 
 }

--- a/core/src/test/java/org/elasticsearch/action/termvectors/GetTermVectorsIT.java
+++ b/core/src/test/java/org/elasticsearch/action/termvectors/GetTermVectorsIT.java
@@ -848,6 +848,16 @@ public class GetTermVectorsIT extends AbstractTermVectorsTestCase {
                 .get();
         assertThat(resp.isExists(), equalTo(true));
         checkBrownFoxTermVector(resp.getFields(), "field1", false);
+
+        // Since the index is empty, all of artificial document's "term_statistics" should be 0/absent
+        Terms terms = resp.getFields().terms("field1");
+        assertEquals("sumDocFreq should be 0 for a non-existing field!", 0, terms.getSumDocFreq());
+        assertEquals("sumTotalTermFreq should be 0 for a non-existing field!", 0, terms.getSumTotalTermFreq());
+        TermsEnum termsEnum = terms.iterator(); // we're guaranteed to receive terms for that field
+        while (termsEnum.next() != null) {
+            String term = termsEnum.term().utf8ToString();
+            assertEquals("term [" + term + "] does not exist in the index; ttf should be 0!", 0, termsEnum.totalTermFreq());
+        }
     }
 
     public void testPerFieldAnalyzer() throws IOException {

